File: rust/dynamic.rs
// @expect verified

trait Tailed {
    fn num_tails(&self) -> isize;
}

struct Fish;
struct TwinTailBeta;

impl Tailed for Fish {
    fn num_tails(&self) -> isize {
        1
    }
}

impl Tailed for TwinTailBeta {
    fn num_tails(&self) -> isize {
        2
    }
}

fn main() {
    let fish: &Tailed = &Fish;
    let beta: &Tailed = &TwinTailBeta;
    assert!(fish.num_tails() == 1);
    assert!(beta.num_tails() == 2);
}

File: kotlin/dynamic.kt
// @expect verified

import assert.*

open class Fish {
    open fun numTails(): Int {
        return 1
    }
}

class TwinTailBeta : Fish() {
    override fun numTails(): Int {
        return 2
    }
}

fun main(args: Array<String>) {
    val fish = Fish()
    val beta = TwinTailBeta()

    __VERIFIER_assert(fish.numTails() == 1)
    __VERIFIER_assert(beta.numTails() == 2)
}
